buildSrc/src/main/java/org/elasticsearch/gradle/precommit/ValidateRestSpecTask.java

```diff
@@ -0,0 +1,160 @@
+ * to the REST specification.
+ */
+public class ValidateRestSpecTask extends PrecommitTask {
+
+    private static final String SCHEMA_PROJECT = ":rest-api-spec";
+    private static final String DOUBLE_STARS = "**"; // checkstyle thinks this is an empty javadoc statement, so string concat instead
+    private static final String JSON_SPEC_PATTERN_INCLUDE = DOUBLE_STARS + "/rest-api-spec/api/" + DOUBLE_STARS + "/*.json";
+    private static final String JSON_SPEC_PATTERN_EXCLUDE = DOUBLE_STARS + "/_common.json";
+    private static final String JSON_SCHEMA_PATTERN = DOUBLE_STARS + "/schema.json";
+    private static final ObjectMapper mapper = new ObjectMapper();
+    private Set<String> ignore = new HashSet<>();
+
+    @SkipWhenEmpty
+    @InputFiles
+    public FileTree getInputDir() {
+        // if null results in in NO-SOURCE
+        return Util.getJavaTestAndMainSourceResources(
+            getProject(),
+            Util.filterByPatterns(
+                getPatternSetFactory().create(),
+                Collections.singletonList(JSON_SPEC_PATTERN_INCLUDE),
+                Collections.singletonList(JSON_SPEC_PATTERN_EXCLUDE)
+            )
+        );
+    }
+
+    @Input
+    @Optional
+    public Set<String> getIgnore() {
+        return ignore;
+    }
+
+    public void ignore(String... ignore) {
+        this.ignore.addAll(Arrays.asList(ignore));
+    }
+
+    @Inject
+    protected Factory<PatternSet> getPatternSetFactory() {
+        throw new UnsupportedOperationException();
+    }
+
+    @OutputFile
+    public File getSuccessMarker() {
+        return new File(getProject().getBuildDir(), "markers/" + this.getName());
+    }
+
+    @TaskAction
+    public void validate() throws IOException {
+        // load schema
+        JsonSchema jsonSchema;
+        FileTree jsonSchemas = Util.getJavaMainSourceResources(
+            getProject().findProject(SCHEMA_PROJECT),
+            Util.filterByPatterns(getPatternSetFactory().create(), Collections.singletonList(JSON_SCHEMA_PATTERN), null)
+        );
+        if (jsonSchemas == null || jsonSchemas.getFiles().size() != 1) {
+            throw new IllegalStateException(
+                String.format(
+                    "Could not find the schema file from glob pattern [%s] and project [%s] for JSON spec validation",
+                    JSON_SCHEMA_PATTERN,
+                    SCHEMA_PROJECT
+                )
+            );
+        }
+        File jsonSchemaOnDisk = jsonSchemas.iterator().next();
+        getLogger().debug("JSON schema for REST spec: [{}]", jsonSchemaOnDisk.getAbsolutePath());
+        SchemaValidatorsConfig config = new SchemaValidatorsConfig();
+        JsonSchemaFactory factory = JsonSchemaFactory.getInstance(SpecVersion.VersionFlag.V7);
+        jsonSchema = factory.getSchema(mapper.readTree(jsonSchemaOnDisk), config);
```

Comment:
I wonder if it's worth doing any kind of caching here, or if the schema stuff does this internally? We want to avoid parsing the schema over and over if possible.

Reply:
I added a simple in-memory cache. not sure how much it helps, but can't hurt.